presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -711,116 +715,136 @@ public static boolean canCoerce(List<? extends Type> actualTypes, List<Type> exp
-                return Optional.empty();
-            }
-            superType = commonSuperType.get();
-        }
-        return Optional.of(superType);
+        Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperType(
+                types.stream()
+                        .map(Type::getTypeSignature)
+                        .collect(toImmutableList()));
+        return commonSuperTypeSignature.map(typeManager::getType);
     }
 
-    public static Optional<Type> getCommonSuperType(Type firstType, Type secondType)
+    public static Optional<Type> getCommonSuperType(Type firstType, Type secondType, TypeManager typeManager)
     {
-        if (firstType.equals(UNKNOWN)) {
-            return Optional.of(secondType);
-        }
+        Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperType(firstType.getTypeSignature(), secondType.getTypeSignature());
+        return commonSuperTypeSignature.map(typeManager::getType);
+    }
 
-        if (secondType.equals(UNKNOWN)) {
-            return Optional.of(firstType);
+    private static Optional<String> getCommonSuperTypeBase(String firstTypeBase, String secondTypeBase)
+    {
+        if (canCastTypeBase(firstTypeBase, secondTypeBase)) {
+            return Optional.of(secondTypeBase);
         }
-
-        if (firstType.equals(secondType)) {
-            return Optional.of(firstType);
+        if (canCastTypeBase(secondTypeBase, firstTypeBase)) {
+            return Optional.of(firstTypeBase);
         }
+        return Optional.empty();
+    }
 
-        if ((firstType.equals(BIGINT) || firstType.equals(DOUBLE)) && (secondType.equals(BIGINT) || secondType.equals(DOUBLE))) {
-            return Optional.<Type>of(DOUBLE);
+    public static Optional<TypeSignature> getCommonSuperType(List<? extends TypeSignature> typeSignatures)
+    {
+        checkArgument(!typeSignatures.isEmpty(), "typeSignatures is empty");
+        TypeSignature superTypeSignature = UNKNOWN.getTypeSignature();
+        for (TypeSignature typeSignature : typeSignatures) {
+            Optional<TypeSignature> commonSuperTypeSignature = getCommonSuperType(superTypeSignature, typeSignature);
+            if (!commonSuperTypeSignature.isPresent()) {
+                return Optional.empty();
+            }
+            superTypeSignature = commonSuperTypeSignature.get();
         }
+        return Optional.of(superTypeSignature);
+    }
 
-        if ((firstType.equals(DATE) || firstType.equals(TIMESTAMP)) && (secondType.equals(DATE) || secondType.equals(TIMESTAMP))) {
-            return Optional.<Type>of(TIMESTAMP);
+    public static Optional<TypeSignature> getCommonSuperType(TypeSignature firstType, TypeSignature secondType)
+    {
+        // Special handling for UnknownType is necessary because we forbid cast between types with different number of type parameters.
+        // Without this, cast between map<bigint, bigint> and unknown will not be allowed.
```

Comment:
Technically speaking, `UNKNOWN` is not a type, so it doesn't make sense say you can cast to it.